src/selector_gui.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file selector_gui.h Functions/types etc. related to the selector widget. */
+
+#ifndef SELECTOR_GUI_H
+#define SELECTOR_GUI_H
+
+#include "stdafx.h"
+#include "querystring_gui.h"
+#include "stringfilter_type.h"
+#include "strings_type.h"
+#include "widget_type.h"
+#include "window_gui.h"
+#include "window_type.h"
+#include "safeguards.h"
+
+struct SelectorWidget {
+
+	std::optional<int> selected_id; ///< ID of the currently selected item
+	std::vector<bool> shown; ///< A vector that determines which items are shown e.g on the graph. Not to be confused with filtered_list
+	std::vector<uint32_t> list; ///< A list of all the items
+	StringFilter string_filter; ///< The filter that checks wether an item should be displayed based on the editbox
+	uint widget_offset; ///< An offset in the parent's window's enum of widgets
+
+	/** A list of items displayed in WID_SELECTION_MATRIX after the editbox filtering
+	 * It is a strict subset of SelectorWidget::list */
+	std::vector<uint32_t> filtered_list;
+
+	static std::unique_ptr<NWidgetBase> MakeSelectorWidgetUI();
+
+	void Init(Window *w);
+	void OnClick(Point pt, WidgetID widget, int click_count);
+	void OnMouseOver(Point pt, WidgetID widget);
+	void OnInvalidateData(int data, bool gui_scope);
+	void UpdateWidgetSize(WidgetID widget, Dimension &size, const Dimension &padding, Dimension &fill, Dimension &resize);
+	void OnResize();
+	void OnEditboxChanged(WidgetID wid);
+	void DrawWidget(const Rect &r, WidgetID widget);
+	void RebuildList();
+
+	virtual ~SelectorWidget() {};
+
+private:
+	Window *w; ///< The parent window pointer
```

Comment:
I don't think `w` is an appropriate name here, `Window *parent_window` would make it entirely clear what it is everywhere.